File: scripts/fix-corrupt.py
#!/usr/bin/env python3
"""
Fix wallet_items, transactions, goals, goal_history corrupt values.
Lee cada valor corrupto, lo desencripta vía Edge Function, y lo actualiza.
"""
import subprocess, json, re, time, sys

ANON_KEY = "sb_publishable_6vRIpL85wKXSvaEK-vui3w_uRl8nYK-"
FUNC_URL = "https://tjbluywadxsbyqtxviyt.supabase.co/functions/v1/finance-crypto"
CWD = "/Users/davidduarte/Documents/Projects/Evolux"

def run(sql):
    r = subprocess.run(["supabase","db","query","--linked",sql],
        capture_output=True, text=True, cwd=CWD)
    return r.stdout

def get_val(val):
    """Check if value needs decryption. Returns (is_encrypted, decrypted_or_None)."""
    if not isinstance(val, str) or len(val) < 10: return False, None
    if '"iv"' in val or '"value"' in val[:20]:
        return True, None  # needs decryption
    return False, None

def decrypt_via_edge(val):
    """Call Edge Function to decrypt a value."""
    payload = json.dumps({"action":"decrypt","value":val,"expectedType":
        "number" if any(c.isdigit() for c in val[:5]) else "string"})
    try:
        r = subprocess.run(["curl","-s","-X","POST",FUNC_URL,
            "-H",f"Authorization: Bearer {ANON_KEY}",
            "-H","Content-Type: application/json",
            "-d",payload], capture_output=True, text=True, timeout=30)
        data = json.loads(r.stdout)
        if "error" not in data and "value" in data:
            return data["value"]
        print(f"      ⚠️  Edge error: {data}")
        return None
    except Exception as e:
        print(f"      ⚠️  Exception: {e}")
        return None

def esc(val, is_num):
    if is_num:
        try: return str(float(val.replace(",",".")))
        except: return "'"+val.replace("'","''")+"'"
    else:
        return "'"+str(val).replace("'","''")+"'"

def is_encrypted(val):
    if not isinstance(val, str) or len(val) < 10: return False
    # The corrupt format stores escaped quotes: {\"iv\":\"...\",\"cipherText\":\"...\"}
    if '\\"iv\\"' in val and '\\"cipherText\\"' in val: return True
    if '\\"iv\\"' in val: return True
    if '"iv"' in val and '"cipherText"' in val: return True
    return False

# Tables to fix: name -> [(field, is_numeric)]
FIX = [
    ("wallet_items", [("name", False), ("value", True)]),
    ("transactions", [("name", False), ("amount", True), ("category", False)]),
    ("goals", [("title", False), ("target", True), ("current", True)]),
    ("goal_history", [("amount", True), ("note", False)]),
]

total = 0
for (table, fields) in FIX:
    print(f"\n━━━ {table}")
    all_cols = ["id"] + [f[0] for f in fields]
    cols = ", ".join(f"'{f}'::text, \"{f}\"::text" for f in all_cols)
    out = run(f"SELECT json_agg(json_build_object({cols})::text) FROM \"{table}\";")
    m = re.search(r'(\[.*\])', out, re.DOTALL)
    if not m:
        print("   No data")
        continue
    raw = re.sub(r'}(\s*){', r'},\1{', m.group(1))
    try: rows = json.loads(raw)
    except:
        print("   Parse error")
        continue
    rows = [json.loads(r) if isinstance(r, str) else r for r in rows]
    print(f"   {len(rows)} rows read")

    for row in rows:
        rid = row.get("id","")
        if not rid: continue
        sets = []
        for fn, is_num in fields:
            val = row.get(fn)
            if is_encrypted(val):
                dec = decrypt_via_edge(val)
                if dec is not None:
                    sets.append(f"\"{fn}\"={esc(dec, is_num)}")
        if sets:
            sql = f"UPDATE \"{table}\" SET {', '.join(sets)} WHERE id='{rid}';"
            o = run(sql)
            if "ERROR" not in o.upper():
                print(f"   ✔ {rid[:12]} → {len(sets)} campos")
                total += 1
            else:
                print(f"   ❌ {rid[:12]} → ERROR: {o[:100]}")
        time.sleep(0.05)

print(f"\n✅ {total} filas actualizadas")

File: scripts/fix-corrupt.mjs
#!/usr/bin/env node
/**
 * Fix corrupt encrypted values in wallet_items, transactions, goals, goal_history.
 * Uses supabase-js for Edge Function calls (which works with anon key)
 * and supabase db query for direct DB access (bypasses RLS).
 */
import { createClient } from '@supabase/supabase-js';
import { execSync } from 'child_process';

const SUPABASE_URL = 'https://tjbluywadxsbyqtxviyt.supabase.co';
const ANON_KEY = 'sb_publishable_6vRIpL85wKXSvaEK-vui3w_uRl8nYK-';
const supabase = createClient(SUPABASE_URL, ANON_KEY);
const CWD = '/Users/davidduarte/Documents/Projects/Evolux';

function run(sql) {
  return execSync(`supabase db query --linked ${JSON.stringify(sql)}`, { cwd: CWD, encoding: 'utf-8' });
}

function getRows(table, fields) {
  const cols = fields.map(f => `'${f}'::text, "${f}"::text`).join(', ');
  const sql = `SELECT json_agg(json_build_object(${cols})::text) FROM "${table}";`;
  const out = run(sql);
  const m = out.match(/(\[.*\])/s);
  if (!m) return [];
  let raw = m[1].replace(/}(\s*){/g, '},$1{');
  try {
    const items = JSON.parse(raw);
    return items.map(i => typeof i === 'string' ? JSON.parse(i) : i);
  } catch { return []; }
}

// Tables to fix: [name, [field, isNumber]]
const FIX = [
  ['wallet_items', [['name', false], ['value', true]]],
  ['transactions', [['name', false], ['amount', true], ['category', false]]],
  ['goals', [['title', false], ['target', true], ['current', true]]],
  ['goal_history', [['amount', true], ['note', false]]],
];

function isEncrypted(val) {
  if (typeof val !== 'string' || val.length < 10) return false;
  // Check both raw and escaped quote formats
  return val.includes('\\"iv\\"') || (val.includes('"iv"') && val.includes('"cipherText"'));
}

function esc(val, isNum) {
  if (isNum) return String(parseFloat(val) || 0);
  return `'${String(val).replace(/'/g, "''")}'`;
}

let total = 0;

for (const [table, fields] of FIX) {
  console.log(`\n━━━ ${table}`);
  const allCols = ['id', ...fields.map(f => f[0])];
  const rows = getRows(table, allCols);
  console.log(`   ${rows.length} rows read`);

  for (const row of rows) {
    const id = row.id;
    if (!id) continue;
    const sets = [];
    for (const [field, isNum] of fields) {
      const val = row[field];
      if (isEncrypted(val)) {
        // Decrypt via Edge Function
        const { data, error } = await supabase.functions.invoke('finance-crypto', {
          body: { action: 'decrypt', value: val, expectedType: isNum ? 'number' : 'string' }
        });
        if (!error && data?.value !== undefined && data.value !== val) {
          sets.push(`"${field}"=${esc(data.value, isNum)}`);
        }
      }
    }
    if (sets.length > 0) {
      const sql = `UPDATE "${table}" SET ${sets.join(', ')} WHERE id='${id}';`;
      run(sql);
      console.log(`   ✔ ${id.slice(0,12)} → ${sets.length} campos`);
      total++;
    }
  }
}

console.log(`\n✅ ${total} filas actualizadas`);
